Multiple buildkite agent support

flokli opened the pull request:
###### Motivation for this change
This is a follow-up of #59358. On top of what's in there, I adressed the requested changes, got rid of `PermissionsStartOnly` and `dataDir`, and added a NixOS VM test testing the ssh key setup (I can't test connecting to buildkite itself, obviously).

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @zimbatm @adisbladis @Lucus16 @rvl @kirelagin

Commit:
buildkite-agent: change option meta-data into tags attrset

yorickvP committed:
services/buildkite-agent: update service for buildkite 3

lukateras committed:
buildkite-agent: make ssh key optional, remove public key option

yorickvP committed:
buildkite-agent2: drop

yorickvP committed:
nixos/buildkite-agents: add module, remove buildkite-agent module

Lucus16 committed:
buildkite-agents: Improve documentation

flokli committed:
nixos/buildkite-agents: remove PermissionsStartOnly, setup dataDir via StateDirectory

flokli committed:
nixos/buildkite-agents: remove deprecated usage of types.string

flokli committed:
nixos/tests: add buildkite-agents test

flokli committed:
nixos/buildkite-agents: fix documentation, defaultTexts

jonringer commented on nixos/modules/services/continuous-integration/buildkite-agents.nix:
```diff
@@ -173,80 +188,106 @@ in
         after = [ "network.target" ];
-        path = cfg.runtimePackages ++ [ pkgs.coreutils ];
+        path = cfg.runtimePackages ++ [ cfg.package pkgs.coreutils ];
+        serviceConfig.StateDirectory = "buildkite-${cfg.userName}";
```
serviceConfig is already defined below on line 270, you should move `StateDirectory = "buildkite-${cfg.userName}";` there

flokli replied:
nice catch! I updated the PR to set it there.

Infinisil commented on nixos/modules/services/continuous-integration/buildkite-agents.nix:
```diff
@@ -68,19 +79,23 @@ in
+            (prev: cur: let pair = builtins.split "=" cur; in
+                 prev // {"${lib.head pair}" = lib.last pair; })
+            {} (lib.remove [] (builtins.split "," commas)); in
+          types.coercedTo types.str commasToAttrs (types.attrsOf types.str);
```
This is rather ugly just to support legacy configuration. I suggest putting this conversion into a `mkChangedOptionModule` instead so this type can be just `attrsOf str`.

flokli replied:
Given we're changing a lot of things in the module anyways, and break backwards compat, we can probably get rid of the old syntax too, WDYT?

Infinisil replied:
Sounds good to me, a `mkRemovedOptionModule` it would be then

flokli replied:
The whole module was changed significantly to allow multiple buildkite agents to be deployed -

I'm not that much a fan of tracking every removed suboption in the imports too - I added a

```
imports = [
    (mkRemovedOptionModule [ "services" "buildkite-agent"] "services.buildkite-agent has been moved to an attribute set at services.buildkite-agents")
  ];
```

do you think that's sufficient?

Infinisil replied:
Yeah that's good

Infinisil commented on nixos/modules/services/continuous-integration/buildkite-agents.nix:
```diff
@@ -43,10 +44,20 @@ in
+        readOnly = true;
+        default = name;
+        description = ''
+          Username of the systemd service this will run as.
```
Maybe use `default = "buildkite-${name}"` instead? I think it would be weird for it to e.g. allocate a user named `foo` just because one uses `foo` as the attribute name.

Infinisil commented on nixos/modules/services/continuous-integration/buildkite-agents.nix:
```diff
@@ -93,26 +108,26 @@ in
+        type = types.nullOr types.path;
+        default = null;
+        description = ''
+          Private agent SSH key.
```
The best way to prevent this from going into the store is to use `toString` in an `apply` directly like
```nix
mkOption {
  apply = final: if final == null then null else toString final;
}
```

Then you don't need to worry about using the result of this option without `toString` (and the `toString` below can be removed).

Infinisil commented on nixos/modules/rename.nix:
```diff
@@ -273,6 +273,9 @@ with lib;
     (mkRenamedOptionModule [ "networking" "resolvconfOptions" ] [ "networking" "resolvconf" "extraOptions" ])
 
+    # Buildkite Agent
+    (mkRemovedOptionModule [ "services" "buildkite-agent"] "services.buildkite-agent has been moved to an attribute set at services.buildkite-agents")
```
See https://github.com/NixOS/nixpkgs/pull/61570, I think these renames should be in the module that needs them

flokli replied:
I don't understand the "that needs them".

`services.buildkite-agent` doesn't exist anymore, you mean I should add those as `imports` to `nixos/modules/services/continuous-integration/buildkite-agents.nix`?

Infinisil replied:
Yeah, like the module that's relevant to the rename/remove

Infinisil commented on pkgs/top-level/all-packages.nix:
```diff
@@ -9248,9 +9248,8 @@ in
   buck = callPackage ../development/tools/build-managers/buck { };
 
-  buildkite-agent = buildkite-agent2;
-  buildkite-agent2 = callPackage ../development/tools/continuous-integration/buildkite-agent/2.x.nix { };
```
I'm a fan of providing at least explanatory errors for removed attributes, something like
```nix
{
  buildkite-agent2 =
    throw "buildkite-agent2 is not supported anymore, use buildkite-agent instead";
}
```

infinisil reviewed (changes requested):
(github wants to have something here)

flokli committed:
nixos/buildkite-agents: remove legacy tag syntax

The module has already changed significantly anyways.

flokli committed:
nixos/buildkite-agents: prefix buildkite- to usernames

It would be weird for it to e.g. allocate a user named foo just because
one uses foo as the attribute name.

flokli committed:
nixos/buildkite-agents: move toString to apply attribute of sshKeyPath

flokli committed:
nixos/buildkite-agents: move mkRemovedOptionModule for old module to imports

flokli committed:
buildkite-agent2: provide `throw` for removed attribute

flokli commented:
Ok, I think I adressed all the remarks. @Infinisil, can you have a second look on it?

flokli commented:
@adisbladis, @zimbatm, can you give this a dry run?

infinisil reviewed (approved):
Code looks good to me, haven't tested it myself though

flokli commented:
All fixes except the multiple buildkite agent support has landed in other PRs by now, and at least I'm fine with only one buildkite agent, so closing.

balsoft commented:
> and at least I'm fine with only one buildkite agent,

:disappointed: 

> closing.

:sob:

flokli commented:
Sorry, this was misunderstanding - I appreciate all the work done here, and PRed most things, it's only that I personally can't pursue the "multiple agents" feature itself. I'd be happy to review a PR by somebody else introducing it :+1: 


Some notes:

As written in https://github.com/NixOS/nixpkgs/pull/78160, handling multiple users and managing secrets is somewhat hard (also see https://github.com/NixOS/rfcs/pull/59).

On top of that, it'd be nice if the service could take use of `DynamicUser=true` and `StateDirectory=buildkite-agent-${n}` (+RuntimeDirectory for configuration file creation).

The current usage of `dataDir` might prohibit this, but if we break API for the multiple buildkite agent feature anyways, we could remove that option too.

nh2 commented:
Concluding comment because PR mentiones do not show up in notifications:

This was implemented in #78373.
